examples/larger-examples/datalog/ford-fulkerson.flix

```diff
@@ -0,0 +1,649 @@
+            Reach(x, y; init(y, x)) :- Edge(x, y, u, f),                 if (u - f) > 0. // Forward edge
+            Reach(x, z; cons(z, p)) :- Reach(x, y; p), Edge(y, z, u, f), if (u - f) > 0. // Forward edge
+            Reach(x, y; init(y, x)) :- Edge(y, x, u, f),                 if f > 0.       // Back edge
+            Reach(x, z; cons(z, p)) :- Reach(x, y; p), Edge(z, y, u, f), if f > 0.       // Back edge
+        };
+        let s = src.src;
+        let t = dst.dst;
+        let result = query edges, rules select fn from Reach(s, t; fn);
+        Vector.head(result) |> Option.flatMap(toOption)
+
+    ///
+    /// Returns the most constraining capacity of `g` on the `Path` `p`.
+    ///
+    def minCapacity(p: Path[a], g: m[FlowNetwork[a]]): Int32 with Foldable[m], Order[a] =
+        let onPath = (s, d) -> isForwardEdge(src = s, dst = d, p) or isBackEdge(src = s, dst = d, p);
+        let optMin = g |> Foldable.filter(match (s, d, _, _) -> onPath(s, d))
+            |> List.map(match (_, _, u, f) -> u - f)
+            |> List.minimum;
+        match optMin {
+            case Some(u) => u
+            case None    => unreachable!() // This function is only called by `maxFlow` if an augmenting path was found
+        }
+
+    ///
+    /// Returns a new flow network where the edges in `g` on the `Path` `p` has been adjusted by `incr`.
+    ///
+    def increaseFlow(p: Path[a], incr: Flow, g: m[FlowNetwork[a]]): Vector[FlowNetwork[a]] with Foldable[m], Order[a] =
+        g
+        |> Foldable.toVector
+        |> Vector.map(match (s, d, u, f) ->
+            if (isForwardEdge(src = s, dst = d, p))
+                (s, d, u, f + incr)
+            else if (isBackEdge(src = s, dst = d, p))
+                (s, d, u, f - incr)
+            else
+                (s, d, u, f)
+        )
+
+    ///
+    /// Returns true if `src` is an edge pointing to `dst` on the `Path` `p`.
+    ///
+    def isForwardEdge(src: { src = a }, dst: { dst = a }, p: Path[a]): Bool with Eq[a] =
+        match (indexOf(src.src, p), indexOf(dst.dst, p)) { // A path is sorted in reverse order
+            case (Some(si), Some(di)) if di + 1 == si => true
+            case _ => false
+        }
+
+    ///
+    /// Returns true if `dst` is an edge pointing to `src` on the `Path` `p`.
+    ///
+    def isBackEdge(src: { src = a }, dst: { dst = a }, p: Path[a]): Bool with Eq[a] =
+        match (indexOf(src.src, p), indexOf(dst.dst, p)) { // A path is sorted in reverse order
+            case (Some(si), Some(di)) if si + 1 == di => true
+            case _ => false
+        }
+
+    pub enum Path[a] with ToString {
+        case Path(List[a])
+        case Bot // Infinitely long path
+    }
+
+    instance Eq[Path[a]] {
+        pub def eq(x: Path[a], y: Path[a]): Bool = match (x, y) {
+            case (Path(xs), Path(ys)) => List.length(xs) == List.length(ys)
+            case (Bot, _)             => false
+            case (_, Bot)             => false
```

Comment:
Bot should be equal to bot.

By convention cases with Bot also goes first. But how could you know that :)